### PR TITLE
Allow containment in adapter trimming

### PR DESCRIPTION
**Pull request by mourisl**
- Handle the case where one read is contained in another read in adapter trimming.
- to resolve the issue in #88

### CODE REVIEW

File: src/chromap.cc [hunk 1]
```diff
@@ -150,8 +168,19 @@ void Chromap::TrimAdapterForPairedEndRead(uint32_t pair_index,
...
             read2_length - seed_start_position + si * seed_length;
-        read_batch1.TrimSequenceAt(pair_index, overlap_length);
-        read_batch2.TrimSequenceAt(pair_index, overlap_length);
+        // The case that read1 is contained in read2
+        int read2_offset = 0; 
+        if (overlap_length > (int)read1_length) {
+          read2_offset = overlap_length - read1_length;
+          overlap_length = read1_length;
+        }
+        if (read1_length <= read2_length) {
```
Comment (haowenz) [inline src/chromap.cc:177]: We make read1 always the shorter one. Then why would we have two cases here (r1<=r2 and r1>r2)?

File: src/sequence_batch.h [hunk 2]
```diff
@@ -129,14 +129,16 @@ class SequenceBatch {
...
     kseq_t *sequence = sequence_batch_[sequence_index];
-    negative_sequence_batch_[sequence_index].erase(
-        negative_sequence_batch_[sequence_index].begin(),
-        negative_sequence_batch_[sequence_index].begin() + sequence->seq.l -
-            length_after_trim);
-    sequence->seq.l = length_after_trim;
-    sequence->seq.s[sequence->seq.l] = '\0';
-    sequence->qual.l = length_after_trim;
-    sequence->qual.s[sequence->qual.l] = '\0';
+    if (length_after_trim < (int)sequence->seq.l) {
```
Comment (haowenz) [inline src/sequence_batch.h:132]: Why would we have a length after trimming longer than the sequence length? It is for the case when the length after trimming is equal to the sequence length where no trimming is performed?

Comment (haowenz) [inline src/sequence_batch.h:132]: Why would we have a trimming length longer than the sequence length?

File: src/chromap.cc [hunk 3]
```diff
@@ -150,8 +168,19 @@ void Chromap::TrimAdapterForPairedEndRead(uint32_t pair_index,
         // Trim adapters and TODO: fix sequencing errors
         int overlap_length =
             read2_length - seed_start_position + si * seed_length;
-        read_batch1.TrimSequenceAt(pair_index, overlap_length);
-        read_batch2.TrimSequenceAt(pair_index, overlap_length);
+        // The case that read1 is contained in read2
```
Comment (haowenz) [inline src/chromap.cc:171]: Can we add more comment here on how this the contained case works? If I understand correctly, when a read is contained in another, no trimming will be performed? And the condition is when the length after trimming is equal to the sequence length?

File: src/chromap.cc [hunk 4]
```diff
@@ -108,10 +108,25 @@ uint32_t Chromap::LoadPairedEndReadsWithBarcodes(SequenceBatch &read_batch1,
 void Chromap::TrimAdapterForPairedEndRead(uint32_t pair_index,
                                           SequenceBatch &read_batch1,
                                           SequenceBatch &read_batch2) {
-  const char *read1 = read_batch1.GetSequenceAt(pair_index);
-  uint32_t read2_length = read_batch2.GetSequenceLengthAt(pair_index);
-  const std::string &negative_read2 =
+  uint32_t raw_read1_length = read_batch1.GetSequenceLengthAt(pair_index);
```
Comment (haowenz) [inline src/chromap.cc:111]: We can make them const.

Comment (haowenz) [inline src/chromap.cc:111]: We can make them const.

File: src/chromap.cc [hunk 5]
```diff
@@ -108,10 +108,25 @@ uint32_t Chromap::LoadPairedEndReadsWithBarcodes(SequenceBatch &read_batch1,
...
+      read_batch1.GetNegativeSequenceAt(pair_index);
+  const std::string &raw_negative_read2 =
       read_batch2.GetNegativeSequenceAt(pair_index);
+  
+  // In the actual adaptor trimming, we assuem length(read1)<=length(read2)
+  //  so we can have the case that read1 is a subset of read2
+  const char *read1 = raw_read1_length <= raw_read2_length ? raw_read1 : raw_read2;
+  const std::string &negative_read2 = raw_read1_length <= raw_read2_length ? 
+                                    raw_negative_read2 : raw_negative_read1;
+  uint32_t read1_length = raw_read1_length <= raw_read2_length ? raw_read1_length :
```
Comment (haowenz) [inline src/chromap.cc:125]: And this const.

**Review by haowenz — CHANGES_REQUESTED** [review]

File: src/sequence_batch.h (hunk 2 above)
Comment (mourisl) [reply src/sequence_batch.h:132]: Right. Because when containment happens, the length_after_trim will be equal to read length. I can add a comment.

**Comment by haowenz** [comment]
DO NOT merge at the moment. It seems that the sequences are not properly trimmed. Need a quick discussion when we are both available.

File: src/chromap.cc [hunk 6]
```diff
@@ -135,7 +157,9 @@ void Chromap::TrimAdapterForPairedEndRead(uint32_t pair_index,
           break;
         }
       }
-      for (uint32_t i = seed_length; i + seed_start_position < read2_length;
+      // The bases after the seed
+      for (uint32_t i = seed_length; i + seed_start_position < read2_length
+          && read1[i];
```
Comment (haowenz) [inline src/chromap.cc:162]: read1[i] seems to be a bug. It should be si * seed_length + I < read1_length.

**Comment by haowenz** [comment]
After some investigation, we noticed that the current method could be too aggressive sometimes. It trims more than what we thought. But this only happens in some rare case where there are tandem repeats and the effect is not big. So we decide to go with the current solution and may change it to require input adapter sequences from users in the future.

**Pull request merged**